google-ads-examples/src/main/java/com/google/ads/googleads/examples/reporting/ParallelReportDownload.java

```diff
@@ -0,0 +1,256 @@
+    // Creates a single client which can be shared by all threads.
+    // gRPC handles multiplexing parallel requests to the underlying API connection.
+    try (GoogleAdsServiceClient serviceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // IMPORTANT: You should avoid hitting the same customer ID in parallel. There are rate limits
+      // at the customer ID level which are much stricter than limits at the developer token level.
+      // Hitting these limits frequently enough will significantly reduce throughput as the client
+      // library will automatically retry with exponential back-off before failing the request.
+      for (String gaqlQuery : GAQL_QUERY_STRINGS) {
+        // Uses a list of futures to make sure that we wait for this report to complete on all
+        // customer IDs before proceeding. The Future data type here is just for demonstration.
+        List<ListenableFuture<ReportSummary>> futures = new ArrayList<>();
+
+        // Uses the API to retrieve the report for each customer ID.
+        for (Long customerId : customerIds) {
+          // Uses the gRPC asynchronous API to download the reports in parallel. This saves having
+          // to create/manage our own thread pool.
+          ResponseCountingObserver responseObserver = new ResponseCountingObserver(customerId);
+
+          // Starts the report download in a background thread.
+          serviceClient
+              .searchStreamCallable()
+              .call(
+                  SearchGoogleAdsStreamRequest.newBuilder()
+                      .setCustomerId(customerId.toString())
+                      .setQuery(gaqlQuery)
+                      .build(),
+                  responseObserver);
+
+          // Stores a future to retrieve the results.
+          futures.add(responseObserver.asFuture());
+        }
+
+        // Waits for all pending requests to the current set of customer IDs to complete.
+        //
+        // This is a naive implementation for illustrative purposes. It is possible to optimize the
+        // utilization of each customer ID by providing a queue of work (or similar). However, this
+        // would complicate the example code and so is omitted here.
+        List<ReportSummary> results = Futures.allAsList(futures).get();
+
+        System.out.println("Report results for query: " + gaqlQuery);
+        results.forEach(System.out::println);
+      }
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** An observer which records a simple count of the result rows received. */
+  private static class ResponseCountingObserver
+      implements ResponseObserver<SearchGoogleAdsStreamResponse> {
+
+    private final long customerId;
+    private final SettableFuture<ReportSummary> future = SettableFuture.create();
+    private final AtomicLong numResponses = new AtomicLong(0);
+    private StreamController controller;
```

Comment:
Not used?

Reply:
Good call - was testing the failure path, forgot to remove this.